packages/matter.js/src/protocol/interaction/AttributeDataEncoder.ts

```diff
@@ -0,0 +1,249 @@
+    attributeReportsPayload?: AttributeReportPayload[];
+    eventReportsPayload?: EventReportPayload[];
+};
+
+/** Encodes an AttributeReportPayload into a TlvStream (used for TlvAny type). */
+export function encodeAttributePayload(attributePayload: AttributeReportPayload): TlvStream {
+    const { attributeData, attributeStatus } = attributePayload;
+    if (attributeData === undefined) {
+        return TlvAttributeReport.encodeTlv({ attributeStatus });
+    }
+
+    const { path, schema, payload, dataVersion } = attributeData;
+    return TlvAttributeReport.encodeTlv({ attributeData: { path, data: schema.encodeTlv(payload), dataVersion } });
+}
+
+/** Encodes an EventReportPayload into a TlvStream (used for TlvAny type). */
+export function encodeEventPayload(eventPayload: EventReportPayload): TlvStream {
+    const { eventData, eventStatus } = eventPayload;
+    if (eventData === undefined) {
+        return TlvEventReport.encodeTlv({ eventStatus });
+    }
+
+    const {
+        path,
+        schema,
+        payload,
+        eventNumber,
+        deltaEpochTimestamp,
+        epochTimestamp,
+        deltaSystemTimestamp,
+        systemTimestamp,
+        priority,
+    } = eventData;
+    return TlvEventReport.encodeTlv({
+        eventData: {
+            path,
+            data: schema.encodeTlv(payload),
+            priority,
+            systemTimestamp,
+            deltaSystemTimestamp,
+            deltaEpochTimestamp,
+            epochTimestamp,
+            eventNumber,
+        },
+    });
+}
+
+/** Return if an AttributeReportPayload can be chunked or not. */
+export function canAttributePayloadBeChunked(attributePayload: AttributeReportPayload): boolean {
+    const { attributeData } = attributePayload;
+    if (attributeData === undefined) {
+        return false;
+    }
+    const {
+        schema,
+        payload,
+        path: { listIndex },
+    } = attributeData;
+    return schema instanceof ArraySchema && Array.isArray(payload) && payload.length > 0 && listIndex === undefined;
+}
+
+/** Chunk an AttributeReportPayload into multiple AttributeReportPayloads. */
+export function chunkAttributePayload(attributePayload: AttributeReportPayload): AttributeReportPayload[] {
+    const { attributeData } = attributePayload;
+    if (attributeData === undefined) {
+        // Should never happen, but we can make sure it do not break
+        return [attributePayload];
+    }
+    const { schema, path, dataVersion, payload } = attributeData;
+    if (!(schema instanceof ArraySchema) || !Array.isArray(payload)) {
+        // Should never happen, but we can make sure it do not break
+        return [attributePayload];
+    }
+    const chunks = new Array<AttributeReportPayload>();
+    chunks.push({ attributeData: { schema, path: { ...path, listIndex: undefined }, payload: [], dataVersion } });
+    payload.forEach(element => {
+        chunks.push({
```

Comment:
If I understand the spec correctly every chunk will require a network round trip before the next one goes.  I think you said CHIP does it that way but oof seems like a lot of latency.  Is it too difficult to integrate w/ serialization so we can pack messages optimally into the assumed 1280 MTU?  Maybe at least a comment

Reply:
Yes it requires to be answered with "ok". The new logic is at least packaging it better then before. Packaging more would require a lot of logic to decide what's best. Especially when the sizes differ because of array chunking and tag compression and then additional requirements apply. So not sure if "one less roundtrip" is worth that computing effort. But yes could be an further optimization for the future.